openmls/src/messages/tests/test_welcome.rs

```diff
@@ -15,10 +16,196 @@ use rstest_reuse::{self, *};
+
+    alice_group
+        .merge_pending_commit()
+        .expect("error merging pending commit");
+
+    // === Deconstruct the Welcome and manipulate the ciphersuite in GroupInfo ===
+
+    let egs = welcome.secrets[0].clone();
+
+    let group_secrets_bytes = backend
+        .crypto()
+        .hpke_open(
+            ciphersuite.hpke_config(),
+            egs.encrypted_group_secrets(),
+            bob_kpb.private_key().as_slice(),
+            &[],
+            &[],
+        )
+        .expect("Could not decrypt group secrets.");
+    let group_secrets = GroupSecrets::tls_deserialize(&mut group_secrets_bytes.as_slice())
+        .expect("Could not deserialize group secrets.")
+        .config(ciphersuite, ProtocolVersion::Mls10);
+    let joiner_secret = group_secrets.joiner_secret;
+
+    // Prepare the PskSecret
+    let psk_secret = PskSecret::new(ciphersuite, backend, group_secrets.psks.psks())
+        .map_err(|e| match e {
+            PskError::LibraryError(e) => e.into(),
+            PskError::TooManyKeys => WelcomeError::PskTooManyKeys,
+            PskError::KeyNotFound => WelcomeError::PskNotFound,
+        })
+        .expect("Could not create PskSecret.");
+
+    // Create key schedule
+    let key_schedule = KeySchedule::init(ciphersuite, backend, joiner_secret, psk_secret)
+        .expect("Could not create KeySchedule.");
+
+    // Derive welcome key & nonce from the key schedule
+    let (welcome_key, welcome_nonce) = key_schedule
+        .welcome(backend)
+        .expect("Using the key schedule in the wrong state")
+        .derive_welcome_key_nonce(backend)
+        .expect("Could not derive welcome key and nonce.");
+
+    let group_info_bytes = welcome_key
+        .aead_open(backend, welcome.encrypted_group_info(), &[], &welcome_nonce)
+        .expect("Could not decrypt GroupInfo.");
+    let mut group_info = GroupInfo::tls_deserialize(&mut group_info_bytes.as_slice())
+        .expect("Could not deserialize GroupInfo.");
+
+    // Manipulate the ciphersuite in the GroupInfo
+    group_info
+        .payload
+        .group_context
+        .set_ciphersuite(Ciphersuite::MLS_256_DHKEMX448_CHACHA20POLY1305_SHA512_Ed448);
+
+    // === Reconstruct the Welcome message and try to process it ===
+
+    let group_info_bytes = group_info
+        .tls_serialize_detached()
+        .expect("Could not serialize GroupInfo.");
+
+    let encrypted_group_info = welcome_key
+        .aead_seal(backend, &group_info_bytes, &[], &welcome_nonce)
+        .expect("Could not encrypt GroupInfo.");
+
+    welcome.encrypted_group_info = encrypted_group_info.into();
+
+    // Bob tries to join the group
+    let err = MlsGroup::new_from_welcome(
+        backend,
+        &mls_group_config,
+        welcome,
+        Some(alice_group.export_ratchet_tree()),
+    )
+    .expect_err("Error creating group from Welcome");
+
+    assert_eq!(err, WelcomeError::GroupInfoCiphersuiteMismatch);
```

Comment:
If it's not too much effort, it would be great if you could add a happy path in addition to the unhappy path.

Reply:
Added below